lib/tokens/TokenManager.js

```diff
@@ -0,0 +1,115 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+'use strict';
+
+const debug = require('debug')('cloudant:tokens:tokenmanager');
+const cookie = require('cookie');
+const EventEmitter = require('events');
+
+class TokenManager {
+  constructor(client, jar, sessionUrl) {
+    this._client = client;
+    this._jar = jar;
+    this._sessionUrl = sessionUrl;
+
+    this._attemptTokenRenewal = true;
+    this._isTokenRenewing = false;
+
+    this._tokenExchangeEE = new EventEmitter().setMaxListeners(Infinity);
+  }
+
+  _autoRenew() {
+    debug('Auto renewing token now...');
+    this._renew().then((response) => {
+      let maxAgeSecs = cookie.parse(response.headers['set-cookie'][0])['Max-Age'] || 3600000;
+      let delaySecs = maxAgeSecs / 2;
+      debug(`Renewing token in ${delaySecs} seconds.`);
+      setTimeout(this._autoRenew.bind(this), delaySecs * 1000).unref();
+    }).catch((error) => {
+      debug(`Failed to auto renew token - ${error}. Retrying in 60 seconds.`);
+      setTimeout(this._autoRenew.bind(this), 60000).unref();
+    });
+  }
+
+  _getToken(callback) {
+    // ** Method to be implemented by _all_ subclasses of `TokenManager` **
+    throw new Error('Not implemented.');
+  }
+
+  // Renew the token.
+  _renew() {
+    if (!this._isTokenRenewing) {
+      this._isTokenRenewing = true;
+      this._tokenExchangeEE.removeAllListeners();
+      debug('Starting token renewal.');
+      this._getToken((error, response) => {
+        if (error) {
+          this._tokenExchangeEE.emit('error', error, response);
+        } else {
+          this._tokenExchangeEE.emit('success', response);
+          this._attemptTokenRenewal = false;
+        }
+        debug('Finished token renewal.');
+        this._isTokenRenewing = false;
+      });
+    }
+    return new Promise((resolve, reject) => {
+      this._tokenExchangeEE.once('success', resolve);
+      this._tokenExchangeEE.once('error', (error, response) => {
+        error.response = response;
+        reject(error);
+      });
+    });
+  }
+
+  // Getter for `attemptTokenRenewal`.
+  // - `true`  A renewal attempt will be made on the next `renew` request.
+  // - `false` No renewal attempt will be made on the next `renew`
+  //           request. Instead the last good renewal response will be returned
+  //           to the client.
+  get attemptTokenRenewal() {
+    return this._attemptTokenRenewal;
+  }
+
+  // Getter for `isTokenRenewing`.
+  // - `true`  A renewal attempt is in progress.
+  // - `false` There are no in progress renewal attempts.
+  get isTokenRenewing() {
+    return this._isTokenRenewing;
+  }
+
+  // Settter for `attemptTokenRenewal`.
+  set attemptTokenRenewal(newAttemptTokenRenewal) {
+    this._attemptTokenRenewal = newAttemptTokenRenewal;
+  }
+
+  // Renew the token if `attemptTokenRenewal` is `true`. Otherwise this is a
+  // no-op so just resolve the promise.
+  renew() {
```

Comment:
I think this should be re-named to make it more clear where it is called what is going to happen e.g.
renewIfRequired
renewIfNeeded
conditionallyRenew

or whatever